task 4

votafore opened the pull request.

votafore committed:
image autoloading

votafore committed:
autofit gridlayout: ready

votafore committed:
list ready

votafore committed:
fragments: ready

votafore committed:
fragments ready

votafore committed:
fragments: ready (no redundant code)

votafore committed:
task almost done

votafore committed:
task4 - little fixes

giffell commented on app/build.gradle:
```diff
@@ -23,6 +23,9 @@ android {
     }
+
+    // переопределение (добавление) каталога ресурсов
+    sourceSets.main.res.srcDir 'src/main/res-test'
```
А стандартный чем плох?

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -28,80 +38,109 @@
-        imageList.setAdapter(app.getAdapter());
+        mFragmentManager = getSupportFragmentManager();
+
+//        RVAdapter.OnItemClickListener adapterListener = new RVAdapter.OnItemClickListener() {
```
Что это за лишний код?

votafore replied:
надо было убрать.... убрал

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -28,80 +38,109 @@
-    ImageView imgView;
-    Button btn_getTop;
-    Button btn_getNew;
+    private View currentView;
```
Она всегда `null`?

votafore replied:
задумывалось что нет
тут она остаться не должна была... убрал

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -28,80 +38,109 @@
-        app.getAdapter().setListener(new RVAdapter.OnItemClickListener() {
-            @Override
-            public void onClick(View item, int position) {
+        FragmentFullImage f = FragmentFullImage.newInstance();
```
Название какое-то странное

votafore replied:
)
Поправил

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -28,80 +38,109 @@
         });
+
+        mFragmentManager.beginTransaction()
+                .replace(R.id.pages, f, "FULL")
```
Почему не в ресурсах?

votafore replied:
1. указывается 1 раз
2. убрал

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -28,80 +38,109 @@
+
+        mFragmentManager.beginTransaction()
+                .replace(R.id.pages, f, "FULL")
+                .setReorderingAllowed(true)
```
А что это за опция и для чего?

votafore replied:
это оптимизация для shared element

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentFullImage.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public void onDetach() {
+        super.onDetach();
+        activityMain.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
```
Там вроде как есть clearFlags метод

votafore replied:
да, но эффект не тот

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentFullImage.java:
```diff
@@ -0,0 +1,66 @@
+    private ActivityMain activityMain;
+    private App app;
+
+    public FragmentFullImage() {
```
А разве его по умолчанию нету?

votafore replied:
убрал

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentFullImage.java:
```diff
@@ -0,0 +1,66 @@
+import com.votafore.earthporn.R;
+
+/**
+ * @author Vorafore
```
Не добавляй хедеры эти, в них нету смысла

votafore replied:
ок

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentFullImage.java:
```diff
@@ -0,0 +1,66 @@
+public class FragmentFullImage extends Fragment {
+
+    private ActivityMain activityMain;
+    private App app;
```
А зачем тут апликейшн контекст?

votafore replied:
для получения картинки

votafore committed:
task4 - little fixes 2

votafore committed:
delete redundant code

votafore committed:
animation is ready

votafore committed:
task finished

votafore committed:
task finished

votafore committed:
fixes

votafore committed:
conflict resolving

# Conflicts:
#	app/src/main/java/com/votafore/earthporn/fragments/FragmentList.java
#	app/src/main/res/transition/open_image.xml

votafore committed:
task4 finished

Commit:
margin in pages

Commit:
margin in pages

giffell commented on app/src/main/java/com/votafore/earthporn/utils/RVAdapter.java:
```diff
@@ -135,8 +76,8 @@ public int getItemCount() {
-        ImageView img;
-        WeakReference<ImageView> ref;
+        public ImageView img;
+        public WeakReference<ImageView> ref;
```
а это зачем?

votafore replied:
для видимости полей вне адаптера

giffell replied:
я про weakreference

votafore replied:
Если в двух словах - это позволяет правильно отображать загружаемые изображения. 
Если описать более подробно, то:
этот объект позволяет сохранять ссылку на ImageView столько сколько нужно. Это используется для того, что бы в элементы списка загружались **их** картинки (а не левые). Т.е. при срабатывании колбэка, когда изображение загружено, будет выполнена проверка есть ли ImageView в WeakReference. Если есть, значит эта вью еще на экране, если нет - вьюхи на экране нет, устанавливать изображение не надо. WeakReference очищается при вызове RecycleViewHolder в адаптере. Т.е. после этого, даже если изображение загрузилось, устанавливать его некуда. А если передать прямую ссылку на ImageView, то я вижу 2 варианта:
1) есть вероятность не правильной установки изображений
2) получить NullPointerException при установке изображения.

giffell commented on app/src/main/java/com/votafore/earthporn/utils/RVAdapter.java:
```diff
@@ -65,67 +52,21 @@ public ImageItem getImageItem(int index){
-            }
-        });
-
+        View itemView = View.inflate(parent.getContext(), R.layout.list_image_item, null);
```
Лучше использовать `Layoutinflater.fromContext()`

votafore replied:
сделал

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentList.java:
```diff
@@ -0,0 +1,151 @@
+                        .setReorderingAllowed(true)
+                        .addSharedElement(item, ViewCompat.getTransitionName(item))
+                        .addToBackStack(pageGallery.toString())
+                        .commit();
```
Тебе не кажется что работу с навигацией лучше вынести в отдельный класс, который будет использоваться во всем приложении?

votafore replied:
кажется, но я еще не придумал как это сделать.

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentGallery.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    public FragmentGallery() {
+        // Required empty public constructor
```
Так зачем переопределять итак пустой конструктор? Ведь он и без переопределения будет существовать

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentFullImage.java:
```diff
@@ -0,0 +1,69 @@
+
+        setSharedElementEnterTransition(TransitionInflater.from(getContext()).inflateTransition(R.transition.open_image));
+
+        ImageItem item = app.getAdapter().getImageItem(App.selectedIndex);
```
Херня какая-то
Из апликейшн класса мы получаем адаптер потом дергаем метод для получения картинки, и при этом сохраняем значение выбранной картинки в том же App классе, все верно?

Application класс не для того чтобы там что-то хранить

votafore replied:
Нет.
Класс содержит (содержал) адаптер. Тут я получаю адаптер и из него достаю ImageItem который находится в списке. Для этого использую индекс сохраненный в App.selectedIndex.

P.S. в новой версии этого уже не будет.

giffell commented:
Этого не должно быть в App классе, это не то место где нужно что-то грузить

    public void onCreate() {

        super.onCreate();

        earthService = new ServiceEarthPorn();
        handler      = new Handler();
        adapter      = new RVAdapter(getApplicationContext());
        pagerAdapter = new PAdapter();
        listener     = new Callback<ListOfImages>() {
            @Override
            public void onResponse(Call<ListOfImages> call, Response<ListOfImages> response) {
                Log.d("NEW_DATA", "received: onResponse");

                final List<ImageItem> list = new ArrayList<>();

                for (Child child: response.body().getData().getChildren()){
                    ImageItem img = new ImageItem();
                    img.item = child.getData();
                    list.add(img);
                }

                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        adapter.setImages(list);
                        pagerAdapter.setImages(list);
                    }
                });
            }

            @Override
            public void onFailure(Call<ListOfImages> call, Throwable t) {
                Log.d("NEW_DATA", "received: onFailure");
                Toast.makeText(getApplicationContext(), "failure", Toast.LENGTH_SHORT).show();
            }
        };
    }
`

Commit:
new image loading mechanism

votafore committed:
task 5 done

Commit:
attempt 1

Commit:
almost done

Commit:
it's done

votafore committed:
task 6 done

Commit:
new animations

giffell commented on app/build.gradle:
```diff
@@ -20,9 +20,16 @@ android {
         }
     }
-    buildToolsVersion '27.0.3'
+    buildToolsVersion '28.0.0 rc1'
```
разве это стабильная?

votafore replied:
выбрал потому что последняя.
исправлю.

giffell commented on app/build.gradle:
```diff
@@ -41,4 +48,5 @@ dependencies {
     annotationProcessor 'com.github.bumptech.glide:compiler:4.6.1'
     implementation 'com.github.chrisbanes:PhotoView:+'
     implementation 'com.android.support:support-v4:27.1.0'
+    implementation 'com.android.support:design:27.1.0'
```
Хорошим тоном считается выносить значения версий в константы

votafore replied:
Исправлю

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -1,107 +1,154 @@
+    private NavigationView nav_view;
+
+    private Fragment fragmentList    = FragmentList.newInstance();
+    private Fragment fragmentGallery = FragmentGallery.newInstance();
```
В чем смысл держать инстансы фрагмента в Активити?

giffell replied:
Почитай что такое FragmentManager и как получать из него фрагменты

votafore replied:
> В чем смысл держать инстансы фрагмента в Активити?

действительно нет)
убрал

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -1,107 +1,154 @@
+        pages    = findViewById(R.id.pages);
+        drawer   = findViewById(R.id.drawer);
+        nav_view = findViewById(R.id.nav_view);
+        toolbar  = findViewById(R.id.toolbar);
```
Подключи ButterKnife

votafore replied:
либу подключил, вот только вопрос появился: есть ли смысл подключать либу из-за нескольких переменных?

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -1,107 +1,154 @@
+            public void onDrawerSlide(View drawerView, float slideOffset) {
+                super.onDrawerSlide(drawerView, slideOffset);
+
+                pages.setTranslationX(80 * slideOffset);
```
Что за магическое число 80? Непонятно, мб 100500 подойдет лучше?)

votafore replied:
это расстояние на которое сдвигается фрагмент со списком фоток при открытии дровера. Баловство конечно) 
Можно убрать.

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -1,107 +1,154 @@
+
+                if(slideOffset == 0){
+                    isLoading = false;
+                    headImg.setImageBitmap(null);
```
Зачем это?

votafore replied:
всегда думал что надо "обнулять" ImageView по возможности.

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -1,107 +1,154 @@
+                List<ImageItem> list = dataSet.getList();
+
+                if (list.size() == 0)
+                    return;
```
Какие-то непонятные проверки, который делают `return` если и надо что-то такое делать, то обычно объединяют их

votafore replied:
ок, попробую исправить

giffell commented on app/src/main/java/com/votafore/earthporn/ActivityMain.java:
```diff
@@ -1,107 +1,154 @@
+        if (currentFragment == null) {
+            mFragmentManager.beginTransaction()
+                    .add(R.id.pages, fragmentList)
+                    .commit();
```
Сделай класс который предоставляет методы для получения текущего фрагмента, удаления, добавления, реплейса фрагментов
Типо вот так
`router.openGalery(fragment, otherParamsIfNeed)`

giffell replied:
А то куча вызов одноразовых по сути

giffell commented on app/src/main/java/com/votafore/earthporn/customviews/AutofitRecyclerView.java:
```diff
@@ -0,0 +1,80 @@
+        super.onMeasure(widthSpec, heightSpec);
+        if (columnWidth > 0) {
+            spanCount = Math.max(1, getMeasuredWidth() / columnWidth);
+            spanCount *= 2;
```
Зачем на две строчки разделять операции для одной и той же переменной?

votafore replied:
исправил

giffell commented on app/src/main/java/com/votafore/earthporn/fragments/FragmentGallery.java:
```diff
@@ -0,0 +1,81 @@
+        ViewPager pager = v.findViewById(R.id.gallery_pager);
+        pager.setAdapter(adapter);
+        pager.setCurrentItem(ActivityMain.selectedIndex);
+        pager.setOnPageChangeListener(new ViewPager.OnPageChangeListener() {
```
добавь в проект retrolambda библиотечку, очень полезна

votafore replied:
на сколько я знаю я и так могу использовать лямбда выражения. Она пригодится дальше?

giffell commented on app/src/main/java/com/votafore/earthporn/models/ImageItem.java:
```diff
@@ -37,8 +31,8 @@ public void setImageToImageView(Context context, final WeakReference<ImageView>
-        int max_width  = 1300;
-        int max_Height = 1200;
+        int max_width  = 1000;
+        int max_Height = 1000;
```
Что за название такое через `_` ? Это не по феншую вообще
И второе, если константа, то пусть будет ее
И в-третьих что за магические числа?

votafore replied:
названия поменял
эти числа указывают максимальный размер изображения:
при портретной ориентации - это макс. высота
при ландшафтной - макс. ширина
Пропорции сохраняются.

giffell commented on app/src/main/java/com/votafore/earthporn/utils/DataSet.java:
```diff
@@ -0,0 +1,52 @@
+        instance.addObserver(observer);
+    }
+
+    public static void removeObserver(Observer observer){
```
Почему методы статические?

votafore replied:
для сокращения вызова. 
Можно, конечно, было сделать и так:
DataSet.getInstance().registerObserver()

giffell commented on app/src/main/java/com/votafore/earthporn/utils/DataSet.java:
```diff
@@ -0,0 +1,52 @@
+
+    /************** data ******************/
+
+    List<ImageItem> list;
```
У тебя ниже есть getter для этого поля, но при этом можно и без него напрямую обратиться к этому полю

votafore replied:
поправлю

giffell commented on app/src/main/java/com/votafore/earthporn/utils/PAdapter.java:
```diff
@@ -50,24 +54,51 @@ public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
     @Override
     public Object instantiateItem(@NonNull ViewGroup container, int position) {
 
+        Log.d("TESTS", String.format("PAdapter: instantiateItem... position:%d", position));
```
Таких логов не должно быть в проде

votafore replied:
исправлю

giffell commented on app/src/main/java/com/votafore/earthporn/utils/RVAdapter.java:
```diff
@@ -1,150 +1,83 @@
-
-    private static final int TYPE_FULL = 0;
-    private static final int TYPE_HALF = 1;
+public class RVAdapter extends RecyclerView.Adapter<RVAdapter.ViewHolder> implements Observer, LifecycleObserver{
```
Название класса не оч, плюс нарушен Java Convention

votafore replied:
поправлю

giffell commented on app/src/main/java/com/votafore/earthporn/utils/RVAdapter.java:
```diff
@@ -1,150 +1,83 @@
-
-    public RVAdapter(Context context){
-        this.context = context;
+        void onClick(int position);
```
Мб уже и не нужен этот интерфейс

Commit:
task 6 completed

Commit:
task 6 to dev

votafore committed:
version 1

votafore committed:
fixes done

votafore committed:
task6 and little fixes

votafore committed:
task6 fragment router ready

votafore committed:
task6 fragment router ready

votafore committed:
resolve conflict

votafore committed:
little fix
